In [43]:
install.packages("car")

also installing the dependencies ‘colorspace’, ‘fracdiff’, ‘lmtest’, ‘urca’, ‘zoo’, ‘RcppArmadillo’, ‘cowplot’, ‘Deriv’, ‘forecast’, ‘modelr’, ‘microbenchmark’, ‘rbibutils’, ‘broom’, ‘doBy’, ‘SparseM’, ‘MatrixModels’, ‘Rdpack’, ‘minqa’, ‘nloptr’, ‘reformulas’, ‘carData’, ‘abind’, ‘Formula’, ‘pbkrtest’, ‘quantreg’, ‘lme4’





The downloaded binary packages are in
	/var/folders/6_/4v4px4w941x057x3xr2z3sb00000gn/T//RtmppIxoQv/downloaded_packages


In [45]:
library("kernlab")
library("kknn")
library("tidyr")
library("dplyr")
library("ggplot2")
library("plotly")
library("combinat")
library("car")
# Load Data Shared with the credit card applications
set.seed(1234)
data_path<- "/Users/pedrofatecha/GitHub/CodePF/MGT6203DataAnalyticsinbusiness/Data/UsedCars.csv"

Used_car <- read.table(data_path, sep = ",", header = TRUE)

In [46]:
# Data Cleaning and Preprocessing

Model <- lm(Used_car$Price ~ Age + KM + HP + Metallic + Automatic + CC + Doors + Gears + Weight, data= Used_car)

Model.residuals <- resid(Model)
Model.fitted <- fitted(Model)


In [47]:
Price_usedcar <- Used_car[1:10,"Price"]

Price_usedcar

Table_check<-cbind("Y"=Price_usedcar[1:10],"Y-hat"=as.vector(Model.fitted)[1:10],"Residuals"=as.vector(Model.residuals)[1:10])

Table_check<-as.data.frame(Table_check)

Table_check[,"Check"]<- Table_check[,"Y"] - Table_check[,"Y-hat"]

Table_check



[1] 21500 20950 19950 19600 21500 22500 22000 22750 17950 16750

Y,Y-hat,Residuals,Check
<dbl>,<dbl>,<dbl>,<dbl>
21500,19186.61,2313.3889,2313.3889
20950,19902.38,1047.6167,1047.6167
19950,20114.27,-164.2689,-164.2689
19600,19891.72,-291.7201,-291.7201
21500,19223.83,2276.1696,2276.1696
22500,18930.85,3569.1535,3569.1535
22000,19698.89,2301.1095,2301.1095
22750,18564.51,4185.4929,4185.4929
17950,16239.52,1710.4842,1710.4842


In [48]:
sum_detail<-summary(Model)

sum_detail


Call:
lm(formula = Used_car$Price ~ Age + KM + HP + Metallic + Automatic + 
    CC + Doors + Gears + Weight, data = Used_car)

Residuals:
    Min      1Q  Median      3Q     Max 
-8117.5  -771.5   -50.0   701.0  6403.5 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -2.883e+03  1.505e+03  -1.916 0.055649 .  
Age         -1.299e+02  2.610e+00 -49.782  < 2e-16 ***
KM          -1.465e-02  1.441e-03 -10.162  < 2e-16 ***
HP           2.540e+01  3.376e+00   7.524 1.01e-13 ***
Metallic    -2.281e+01  7.436e+01  -0.307 0.759059    
Automatic    4.991e+02  1.497e+02   3.334 0.000881 ***
CC          -5.259e-03  8.499e-02  -0.062 0.950667    
Doors        1.315e+01  4.086e+01   0.322 0.747638    
Gears        6.281e+02  1.927e+02   3.259 0.001148 ** 
Weight       1.504e+01  1.300e+00  11.567  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1227 on 1254 degrees of freedom
Multiple R-squared:  0.8649,	Adjus

In [49]:
beta_estimate <- sum_detail$coefficients[,1]
beta_estimate

Standard_error <- sum_detail$coefficients[,2]

tstat<- beta_estimate / Standard_error

cbind(tstat,sum_detail$coefficients[,3])



(Intercept)           Age            KM            HP      Metallic 
-2.882537e+03 -1.299451e+02 -1.464727e-02  2.539843e+01 -2.281278e+01 
    Automatic            CC         Doors         Gears        Weight 
 4.990580e+02 -5.259189e-03  1.315082e+01  6.281485e+02  1.503712e+01

,tstat,
(Intercept),-1.91555684,-1.91555684
Age,-49.78213520,-49.78213520
KM,-10.16168388,-10.16168388
HP,7.52407315,7.52407315
Metallic,-0.30678348,-0.30678348
Automatic,3.33400809,3.33400809
CC,-0.06188146,-0.06188146
Doors,0.32182598,0.32182598
Gears,3.25899198,3.25899198
Weight,11.56676201,11.56676201


In [53]:
# T statistic
Degrees_freedom <- Model$df.residual
Degrees_freedom
alpha<- 1- 0.95

1-(alpha/2)

qt(1-(alpha/2), Degrees_freedom)

qnorm(1-(alpha/2))



[1] 1254

[1] 0.975

[1] 1.961858

[1] 1.959964

In [67]:
pval<- 2 * pt(-abs(tstat), Degrees_freedom)

as.vector(pval)<=0.05

sig_ind<-ifelse(as.vector(pval)<=0.05, "significant","not significant")

cbind(Pvalue=pval,calc_Pvalue=sum_detail$coefficients[,4],tstat=sum_detail$coefficients[,3],significants=sig_ind)

sum_detail$coefficients[,1]/   sum_detail$coefficients[,2]


[1] FALSE  TRUE  TRUE  TRUE FALSE  TRUE FALSE FALSE  TRUE  TRUE

,Pvalue,calc_Pvalue,tstat,significants
(Intercept),0.0556487459896112,0.0556487459896112,-1.91555684361171,not significant
Age,2.80253257179396e-299,2.80253257179396e-299,-49.7821352014176,significant
KM,2.2952362856837e-23,2.2952362856837e-23,-10.1616838774669,significant
HP,1.00919988080613e-13,1.00919988080613e-13,7.52407315256202,significant
Metallic,0.759059117007503,0.759059117007503,-0.306783478398259,not significant
Automatic,0.000881043190494714,0.000881043190494714,3.33400809260806,significant
CC,0.950667091527915,0.950667091527915,-0.0618814620178379,not significant
Doors,0.74763816780059,0.74763816780059,0.321825981098421,not significant
Gears,0.00114807895738543,0.00114807895738543,3.25899197829076,significant
Weight,1.79075894090841e-29,1.79075894090841e-29,11.5667620114143,significant


(Intercept)          Age           KM           HP     Metallic    Automatic 
 -1.91555684 -49.78213520 -10.16168388   7.52407315  -0.30678348   3.33400809 
          CC        Doors        Gears       Weight 
 -0.06188146   0.32182598   3.25899198  11.56676201

### the variables with more significance for the model are AGE, KM,HP, Automatic,Gears,Weight

In [66]:
Model_2_reduced<- lm(Used_car$Price ~ Age + KM + HP + Automatic + Gears + Weight, data= Used_car)
Model_3_reduced<- lm(Used_car$Price ~ Age + KM + HP + Automatic + Weight, data= Used_car)
Model_4_reduced<- lm(Used_car$Price ~ Age + KM + Gears + HP +Automatic, data= Used_car)

Model_2_reduced_summary<- summary(Model_2_reduced)
Model_2_reduced_summary
Model_3_reduced_summary<- summary(Model_3_reduced)
Model_3_reduced_summary
Model_4_reduced_summary<- summary(Model_4_reduced)
Model_4_reduced_summary
y_hat <- fitted(Model_2_reduced)
u_hat <- resid(Model_2_reduced)

Model_vec<- c("Model 1", "Model 2", "Model 3","Model 4")
rsquare_vec <- c(sum_detail$r.squared,Model_2_reduced_summary$r.squared,Model_3_reduced_summary$r.squared,Model_4_reduced_summary$r.squared)
adj_rsquare_vec <- c(sum_detail$adj.r.squared,Model_2_reduced_summary$adj.r.squared,Model_3_reduced_summary$adj.r.squared,Model_4_reduced_summary$adj.r.squared)

cbind(Model=Model_vec, R_squared=rsquare_vec, Adj_R_squared=adj_rsquare_vec)




Call:
lm(formula = Used_car$Price ~ Age + KM + HP + Automatic + Gears + 
    Weight, data = Used_car)

Residuals:
    Min      1Q  Median      3Q     Max 
-8193.8  -772.8   -47.9   703.3  6402.3 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -2.930e+03  1.498e+03  -1.956 0.050666 .  
Age         -1.299e+02  2.600e+00 -49.942  < 2e-16 ***
KM          -1.463e-02  1.439e-03 -10.170  < 2e-16 ***
HP           2.527e+01  3.294e+00   7.671  3.4e-14 ***
Automatic    4.930e+02  1.476e+02   3.341 0.000861 ***
Gears        6.139e+02  1.876e+02   3.272 0.001098 ** 
Weight       1.519e+01  1.202e+00  12.637  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1226 on 1257 degrees of freedom
Multiple R-squared:  0.8648,	Adjusted R-squared:  0.8642 
F-statistic:  1341 on 6 and 1257 DF,  p-value: < 2.2e-16



Call:
lm(formula = Used_car$Price ~ Age + KM + HP + Automatic + Weight, 
    data = Used_car)

Residuals:
    Min      1Q  Median      3Q     Max 
-8185.8  -777.1   -40.8   694.0  6380.3 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -1.318e+01  1.208e+03  -0.011  0.99130    
Age         -1.299e+02  2.610e+00 -49.759  < 2e-16 ***
KM          -1.444e-02  1.443e-03 -10.010  < 2e-16 ***
HP           2.766e+01  3.224e+00   8.579  < 2e-16 ***
Automatic    4.646e+02  1.479e+02   3.142  0.00172 ** 
Weight       1.510e+01  1.206e+00  12.523  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1231 on 1258 degrees of freedom
Multiple R-squared:  0.8637,	Adjusted R-squared:  0.8632 
F-statistic:  1594 on 5 and 1258 DF,  p-value: < 2.2e-16



Call:
lm(formula = Used_car$Price ~ Age + KM + Gears + HP + Automatic, 
    data = Used_car)

Residuals:
    Min      1Q  Median      3Q     Max 
-4011.6  -866.1   -35.0   756.9  6812.4 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  1.193e+04  9.842e+02  12.123  < 2e-16 ***
Age         -1.425e+02  2.548e+00 -55.912  < 2e-16 ***
KM          -1.435e-02  1.526e-03  -9.401  < 2e-16 ***
Gears        5.648e+02  1.991e+02   2.837  0.00462 ** 
HP           4.614e+01  3.025e+00  15.253  < 2e-16 ***
Automatic    9.316e+02  1.522e+02   6.120 1.25e-09 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1301 on 1258 degrees of freedom
Multiple R-squared:  0.8477,	Adjusted R-squared:  0.8471 
F-statistic:  1400 on 5 and 1258 DF,  p-value: < 2.2e-16


Model,R_squared,Adj_R_squared
Model 1,0.864867726797135,0.863897877946397
Model 2,0.86484736620537,0.864202246234989
Model 3,0.86369644599006,0.863154698955044
Model 4,0.847678407370533,0.847072995635121


In [52]:
# r-squared

R_squared <- sum_detail$r.squared
R_squared

# ess/tss
var(Model.fitted)/var(Used_car$Price)

# rss/tss

1- var(Model.residuals)/var(Used_car$Price)

# Colinearity

vif(Model)


[1] 0.8648677

[1] 0.8648677

[1] 0.8648677

Age        KM        HP  Metallic Automatic        CC     Doors     Gears 
 1.914772  1.586299  1.548521  1.017014  1.101810  1.104307  1.269749  1.129618 
   Weight 
 2.096541

In [68]:
format(9.506671e-01, scientific = FALSE)

[1] "0.9506671"

In [71]:
Model_weight <- lm(Used_car$Weight ~ Age + KM + HP + Metallic + Automatic + CC + Doors + Gears, data= Used_car)

summary(Model_weight)

vif(Model_weight)



Call:
lm(formula = Used_car$Weight ~ Age + KM + HP + Metallic + Automatic + 
    CC + Doors + Gears, data = Used_car)

Residuals:
   Min     1Q Median     3Q    Max 
-62.22 -12.80  -0.52   7.99 527.97 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  8.739e+02  2.143e+01  40.784   <2e-16 ***
Age         -7.322e-01  5.278e-02 -13.873   <2e-16 ***
KM          -1.229e-06  3.130e-05  -0.039   0.9687    
HP           1.210e+00  6.486e-02  18.651   <2e-16 ***
Metallic    -1.061e+00  1.614e+00  -0.657   0.5112    
Automatic    2.953e+01  3.142e+00   9.399   <2e-16 ***
CC           3.710e-03  1.842e-03   2.014   0.0443 *  
Doors        1.174e+01  8.231e-01  14.261   <2e-16 ***
Gears        9.766e+00  4.176e+00   2.339   0.0195 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 26.65 on 1255 degrees of freedom
Multiple R-squared:  0.523,	Adjusted R-squared:   0.52 
F-statistic:   172 on 8 and 1255 DF,  p-value: < 2

Age        KM        HP  Metallic Automatic        CC     Doors     Gears 
 1.660166  1.586297  1.212444  1.016664  1.029355  1.100751  1.092669  1.124717

In [73]:
Model_2_reduced$coefficients["Age"]*12

Age 
-1558.463

In [74]:
Model_2_reduced$coefficients["KM"]*10000

KM 
-146.3109

# Part 2 Homework 1

In [95]:
path_data="/Users/pedrofatecha/GitHub/CodePF/MGT6203DataAnalyticsinbusiness/Data/Loan.csv"

Loan <- read.table(path_data, sep = ",", header = TRUE)

Loan[1:10,]

,Loan,Income,Family,CCAvg,Education
,<int>,<int>,<int>,<dbl>,<int>
1,0,49,4,1.6,1
2,0,34,3,1.5,1
3,0,11,1,1.0,1
4,0,100,1,2.7,2
5,0,45,4,1.0,2
6,0,29,4,0.4,2
7,0,72,2,1.5,2
8,0,22,1,0.3,3
9,0,81,3,0.6,2


In [96]:
str(Loan)

'data.frame':	5000 obs. of  5 variables:
 $ Loan     : int  0 0 0 0 0 0 0 0 0 1 ...
 $ Income   : int  49 34 11 100 45 29 72 22 81 180 ...
 $ Family   : int  4 3 1 1 4 4 2 1 3 1 ...
 $ CCAvg    : num  1.6 1.5 1 2.7 1 0.4 1.5 0.3 0.6 8.9 ...
 $ Education: int  1 1 1 2 2 2 2 3 2 3 ...


In [161]:
Loan$Education<-as.factor(Loan$Education)

Linear_Model <- lm(Loan ~ Income + Family + CCAvg +Education, data = Loan)


Linear_Model_Res<-summary(logit_model)

Linear_Model_Res
str(Loan)


Call:
glm(formula = Loan ~ Income + Family + CCAvg + Education, family = binomial(link = logit), 
    data = Loan)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -13.177833   0.517777 -25.451  < 2e-16 ***
Income        0.059791   0.002687  22.255  < 2e-16 ***
Family        0.587079   0.071275   8.237  < 2e-16 ***
CCAvg         0.162679   0.040505   4.016 5.91e-05 ***
Education2    3.910609   0.251037  15.578  < 2e-16 ***
Education3    3.933173   0.244329  16.098  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 3162.0  on 4999  degrees of freedom
Residual deviance: 1334.8  on 4994  degrees of freedom
AIC: 1346.8

Number of Fisher Scoring iterations: 8


'data.frame':	5000 obs. of  5 variables:
 $ Loan     : int  0 0 0 0 0 0 0 0 0 1 ...
 $ Income   : int  49 34 11 100 45 29 72 22 81 180 ...
 $ Family   : int  4 3 1 1 4 4 2 1 3 1 ...
 $ CCAvg    : num  1.6 1.5 1 2.7 1 0.4 1.5 0.3 0.6 8.9 ...
 $ Education: Factor w/ 3 levels "1","2","3": 1 1 1 2 2 2 2 3 2 3 ...


In [123]:
Linear_Model$coefficients

(Intercept)       Income       Family        CCAvg   Education2   Education3 
-0.345540917  0.003367256  0.031602457  0.013734629  0.151743228  0.160546743

Interpretation
==========
<link rel="stylesheet" href="/Users/pedrofatecha/GitHub/CodePF/ISYE6501_IntrotoAnalyticsModeling/02_Script/02_1_Function/styles.css">

> The Education variable is categorical with three levels (1=Undergrad, 2=Graduate, 3=Advanced/Professional), using Undergraduate as the reference category.</br></br>The Education2 coefficient (0.1517) indicates that customers with GRADUATE degrees are 15.17 percentage points more likely to accept the personal loan offer compared to customers with Undergraduate degrees, holding all other variables (Income, Family, CCAvg) constant.</br></br>The Education3 coefficient (0.1605) indicates that customers with Advanced/Professional degrees are 16.05 percentage points more likely to accept the loan compared to customers with Undergraduate degrees, holding all other variables constant.</br></br>The difference between Advanced and Graduate degree holders is 0.88 percentage points (0.1605 - 0.1517), meaning Advanced degree holders are slightly more likely to accept than Graduate degree holders. Both education levels show substantially higher loan acceptance rates than the Undergraduate baseline.
</br></br>


#### the fitted model of a linear probability model means is the predicted probability of the variable that I am trying to regress
#### there is not customers greater than one but we have cases with less than zero

In [164]:


#head(sort(predict(Linear_Model)),10)

#head(sort(Linear_Model$fitted.values),10)

#tail(sort(predict(Linear_Model)),10)

Linear_Model$fitted.values[min(Linear_Model$fitted.values)==Linear_Model$fitted.values]


1483 
-0.2856269

In [172]:
logit_model <- glm(Loan ~ Income + Family + CCAvg +Education, data = Loan, family = binomial(link=logit))

logit_model_res <- summary(logit_model)
y_hat_logit <- rep(1,nrow(Loan))

y_hat_logit[logit_model$fitted.values < mean(Loan$Loan)] <- 0

Confusion<- table(y_hat_logit,Loan$Loan)
Confusion
accuracy <- sum(diag(Confusion))/sum(Confusion)
accuracy
cbind(Confusion[2,2]/sum(Confusion[,2]),Confusion[1,1]/sum(Confusion[,1]))



           
y_hat_logit    0    1
          0 4001   61
          1  519  419

[1] 0.884

0.8729167,0.885177


In [173]:
coef(logit_model)

(Intercept)       Income       Family        CCAvg   Education2   Education3 
-13.17783285   0.05979075   0.58707882   0.16267911   3.91060897   3.93317273

In [180]:
coef(logit_model)
colMeans(Loan[,2:4])

predict(logit_model, newdata = as.data.frame(new_data), type="response")

exp(-13.1778328473338 + 0.0597907530912975 *  73.7742 + 0.587078822993878 * 2.3964 + 0.162679106664809 * 1.93793800000001 + 1* 3.91060896525892 + 0 * 3.93317272591235)/(1+exp(-13.1778328473338 + 0.0597907530912975 *  73.7742 + 0.587078822993878 * 2.3964 + 0.162679106664809 * 1.93793800000001 + 1 * 3.91060896525892 + 0 * 3.93317272591235))

(Intercept)       Income       Family        CCAvg   Education2   Education3 
-13.17783285   0.05979075   0.58707882   0.16267911   3.91060897   3.93317273

Income    Family     CCAvg 
73.774200  2.396400  1.937938

ERROR: Error in eval(predvars, data, env): object 'Income' not found


In [166]:
new_data <- data.frame(Income = mean(Loan$Income), Family = mean(Loan$Family), 
                       CCAvg = mean(Loan$CCAvg), Education = factor(2, levels=c(1,2,3)))

predict(logit_model, newdata = new_data, type = "response")
logit_model$coefficients
Coeff_model_logit<-as.vector(logit_model$coefficients)

new_data<- as.vector(c(1, mean(Loan$Income), mean(Loan$Family), mean(Loan$CCAvg), 1, 0))
new_data
Coeff_model_logit[2:length(Coeff_model_logit)]

log_odds <- crossprod(Coeff_model_logit,  new_data)

log_odds

exp(log_odds)/(1+exp(log_odds))


as.vector(Linear_Model$coefficients)%*%new_data

cbind(logistic_model=Coeff_model_logit,Linear_Model=Linear_Model$coefficients)



1 
0.04172353

(Intercept)       Income       Family        CCAvg   Education2   Education3 
-13.17783285   0.05979075   0.58707882   0.16267911   3.91060897   3.93317273

[1]  1.000000 73.774200  2.396400  1.937938  1.000000  0.000000

[1] 0.05979075 0.58707882 0.16267911 3.91060897 3.93317273

-3.134071


0.04172353


0.1569679


,logistic_model,Linear_Model
(Intercept),-13.17783285,-0.345540917
Income,0.05979075,0.003367256
Family,0.58707882,0.031602457
CCAvg,0.16267911,0.013734629
Education2,3.91060897,0.151743228
Education3,3.93317273,0.160546743


In [182]:
# X point where we evaluate marginal effects
new_data <- data.frame(
  Income   = mean(Loan$Income),
  Family   = mean(Loan$Family),
  CCAvg    = mean(Loan$CCAvg),
  Education = factor(2, levels = c(1, 2, 3))
)

# Linear predictor Xβ (log-odds)
xb <- predict(logit_model, newdata = new_data, type = "link")

# Probability at that point
p_hat <- plogis(xb)   # same as predict(..., type="response")

p_hat

1 
0.04172353

In [188]:
lambda_prime <- dlogis(xb)  # derivative of logistic CDF at Xβ
lambda_prime

1 
0.03998267

In [184]:
beta_logit <- coef(logit_model)

# For continuous predictors, partial effect = dlogis(Xβ) * beta_j
pe_logit <- lambda_prime * beta_logit

pe_logit

(Intercept)       Income       Family        CCAvg   Education2   Education3 
-0.526884985  0.002390594  0.023472981  0.006504346  0.156356601  0.157258760

In [185]:
pe_logit_vars <- pe_logit[-1]   # drop intercept
pe_logit_vars

Income      Family       CCAvg  Education2  Education3 
0.002390594 0.023472981 0.006504346 0.156356601 0.157258760

In [186]:
Linear_Model <- lm(Loan ~ Income + Family + CCAvg + Education, data = Loan)

beta_lpm <- coef(Linear_Model)
beta_lpm_vars <- beta_lpm[-1]   # drop intercept
beta_lpm_vars

Income      Family       CCAvg  Education2  Education3 
0.003367256 0.031602457 0.013734629 0.151743228 0.160546743

In [187]:
comparison <- cbind(
  LPM_partial  = beta_lpm_vars,
  Logit_partial = as.numeric(pe_logit_vars)
)
comparison

,LPM_partial,Logit_partial
Income,0.003367256,0.002390594
Family,0.031602457,0.023472981
CCAvg,0.013734629,0.006504346
Education2,0.151743228,0.156356601
Education3,0.160546743,0.157258760
